things/service/src/main/java/org/eclipse/ditto/things/service/utils/PartialAccessPathCalculator.java

```diff
@@ -0,0 +1,338 @@
+
+import org.eclipse.ditto.base.model.auth.AuthorizationContext;
+import org.eclipse.ditto.base.model.auth.AuthorizationSubject;
+import org.eclipse.ditto.base.model.auth.DittoAuthorizationContextType;
+import org.eclipse.ditto.json.JsonArrayBuilder;
+import org.eclipse.ditto.json.JsonFactory;
+import org.eclipse.ditto.json.JsonObject;
+import org.eclipse.ditto.json.JsonObjectBuilder;
+import org.eclipse.ditto.json.JsonPointer;
+import org.eclipse.ditto.policies.api.Permission;
+import org.eclipse.ditto.policies.enforcement.PolicyEnforcer;
+import org.eclipse.ditto.policies.model.Permissions;
+import org.eclipse.ditto.policies.model.PoliciesResourceType;
+import org.eclipse.ditto.policies.model.ResourceKey;
+import org.eclipse.ditto.policies.model.enforcers.Enforcer;
+import org.eclipse.ditto.things.model.Thing;
+
+/**
+ * Utility class for calculating which JSON paths each subject with partial READ permissions can access.
+ * <p>
+ * This class is used to enable partial change notifications for subjects with restricted READ permissions.
+ * It calculates accessible paths for subjects that have partial (restricted) access to a Thing, excluding
+ * subjects that have full access to the root resource.
+ * </p>
+ * <p>
+ * The calculated paths are returned in an indexed format to reduce header size when serialized.
+ * </p>
+ *
+ * @since 3.9.0
+ */
+public final class PartialAccessPathCalculator {
+
+    /**
+     * JSON key for the subjects array in the partial access paths structure.
+     */
+    public static final String SUBJECTS_KEY = "subjects";
+    
+    /**
+     * JSON key for the paths object in the partial access paths structure.
+     */
+    public static final String PATHS_KEY = "paths";
+
+    private static final JsonPointer ROOT_RESOURCE_POINTER = JsonPointer.empty();
+    private static final Permissions READ_PERMISSIONS = Permissions.newInstance(Permission.READ);
+
+    private PartialAccessPathCalculator() {
+        throw new AssertionError("This class is not meant to be instantiated.");
+    }
+
+    /**
+     * Calculates which JSON paths each subject with partial READ permissions can access.
+     * <p>
+     * This method identifies subjects with partial permissions and calculates their accessible paths,
+     * excluding subjects that have full access to the root resource (as they don't need partial paths).
+     * </p>
+     *
+     * @param thing the current Thing state (may be null for ThingDeleted events)
+     * @param policyEnforcer the PolicyEnforcer to use for permission checks (must not be null)
+     * @return a map from subject ID to list of accessible JsonPointer paths, empty map if no partial subjects exist
+     * @throws NullPointerException if policyEnforcer is null
+     */
+    public static Map<String, List<JsonPointer>> calculatePartialAccessPaths(
+            @Nullable final Thing thing,
+            final PolicyEnforcer policyEnforcer) {
+
+        if (thing == null || policyEnforcer == null) {
```

Comment:
`policyEnforcer` cannot be null at this place.
If it can be `null`, the parameter must be declared as `@Nullable`.
However, the only place calling it makes sure it is not null.